File: snow_pkg/src/cs/Preciprec.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class Preciprec
{
    private double _rho;
    public double rho
        {
            get { return this._rho; }
            set { this._rho= value; } 
        }
    public Preciprec() { }
    
    public void  CalculateModel(SnowState s, SnowState s1, SnowRate r, SnowAuxiliary a)
    {
        //- Name: Preciprec -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: Precipitation ReCalculation
    //            * Author: STICS
    //            * Reference: -
    //            * Institution: INRA
    //            * Abstract: -
        //- inputs:
    //            * name: Sdry_t1
    //                          ** description : water in solid state in the snow cover in previous day
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 500.0
    //                          ** unit : mmW
    //                          ** uri : 
    //            * name: Sdry
    //                          ** description : water in solid state in the snow cover 
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 500.0
    //                          ** unit : mmW
    //                          ** uri : 
    //            * name: Swet
    //                          ** description : water in liquid state in the snow cover
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 100.0
    //                          ** unit : mmW
    //                          ** uri : 
    //            * name: Swet_t1
    //                          ** description : water in liquid state in the snow cover in previous day
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 100.0
    //                          ** unit : mmW
    //                          ** uri : 
    //            * name: Sdepth_t1
    //                          ** description : snow cover depth Calculation in previous day
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 5000.0
    //                          ** unit : m
    //                          ** uri : 
    //            * name: Sdepth
    //                          ** description : snow cover depth Calculation
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 5000.0
    //                          ** unit : m
    //                          ** uri : 
    //            * name: Mrf
    //                          ** description : liquid water in the snow cover in the process of refreezing
    //                          ** inputtype : variable
    //                          ** variablecategory : rate
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : mmW/d
    //                          ** uri : 
    //            * name: precip
    //                          ** description : recalculated precipitation
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 5000.0
    //                          ** unit : mmW
    //                          ** uri : 
    //            * name: Snowaccu
    //                          ** description : snowfall accumulation
    //                          ** inputtype : variable
    //                          ** variablecategory : rate
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : mmW/d
    //                          ** uri : 
    //            * name: rho
    //                          ** description : The density of the new snow fixed by the user
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** default : 100
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : kg/m**3
    //                          ** uri : 
        //- outputs:
    //            * name: preciprec
    //                          ** description : precipitation recalculation
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0.0
    //                          ** max : 500.0
    //                          ** unit : mmW
    //                          ** uri : 
        double Sdry_t1 = s1.Sdry;
        double Sdry = s.Sdry;
        double Swet = s.Swet;
        double Swet_t1 = s1.Swet;
        double Sdepth_t1 = s1.Sdepth;
        double Sdepth = s.Sdepth;
        double Mrf = r.Mrf;
        double precip = a.precip;
        double Snowaccu = r.Snowaccu;
        double preciprec;
        preciprec = precip;
        if (Sdry + Swet < Sdry_t1 + Swet_t1)
        {
            preciprec = preciprec + ((Sdepth_t1 - Sdepth) * rho) - Mrf;
        }
        preciprec = preciprec - Snowaccu;
        s.preciprec= preciprec;
    }
}

File: snow_pkg/src/record/Preciprec.cpp
// @@tagdynamic@@
// @@tagdepends: vle.discrete-time @@endtagdepends

#include <vle/DiscreteTime.hpp>
namespace vd = vle::devs;
namespace vv = vle::value;
// Definition du namespace de la classe du modele
namespace record {
namespace Snow {
using namespace vle::discrete_time;
class Preciprec: public DiscreteTimeDyn {
public:
    Preciprec(const vd::DynamicsInit& atom, const vd::InitEventList& events) : DiscreteTimeDyn(atom, events)
    {
        // Ces parametres ont une valeur par defaut utilise si la condition n'est pas definie
        rho = (events.exist("rho")) ? vv::toDouble(events.get("rho")) : 100;
        //  Variables gérées par ce composant
        preciprec.init(this,"preciprec", events);
        //  Variables gérées par un autre composant
        Sdry.init(this,"Sdry", events);
        Swet.init(this,"Swet", events);
        Sdepth.init(this,"Sdepth", events);
        Mrf.init(this,"Mrf", events);
        Snowaccu.init(this,"Snowaccu", events);
        precip.init(this,"precip", events);
    }
    /**
    * @brief Destructeur de la classe du modèle.
    **/
    virtual ~Preciprec() {};
    /**
    * @brief Methode de calcul effectuée à chaque pas de temps
    * @param time la date du pas de temps courant
    */
    virtual void compute(const vd::Time& /*time*/)
    {
        preciprec = precip();
        if (Sdry() + Swet() < Sdry(-1) + Swet(-1))
        {
            preciprec = preciprec() + ((Sdepth(-1) - Sdepth()) * rho) - Mrf();
        }
        preciprec = preciprec() - Snowaccu();
    }
private:
    //Variables d'etat
    /**
    * @brief precipitation recalculation (mmW)
    */
    Var preciprec

    //Entrées
    /**
        * @brief water in solid state in the snow cover in previous day (mmW)
        */
    Var Sdry/**
        * @brief water in liquid state in the snow cover (mmW)
        */
    Var Swet/**
        * @brief snow cover depth Calculation in previous day (m)
        */
    Var Sdepth/**
        * @brief liquid water in the snow cover in the process of refreezing (mmW/d)
        */
    Var Mrf/**
        * @brief snowfall accumulation (mmW/d)
        */
    Var Snowaccu/**
        * @brief recalculated precipitation (mmW)
        */
    Var precip

    //Paramètres du modele
    /**
    * @brief The density of the new snow fixed by the user (kg/m**3)
    */
    double rho;
};
}
}
DECLARE_DYNAMICS(record::Snow::Preciprec); // balise specifique VLE